cores/esp8266/core_esp8266_waveform.cpp

```diff
@@ -112,20 +114,30 @@ void setTimer1Callback(uint32_t (*fn)()) {
 // waveform smoothly on next low->high transition.  For immediate change, stopWaveform()
 // first, then it will immediately begin.
 int startWaveform(uint8_t pin, uint32_t timeHighUS, uint32_t timeLowUS, uint32_t runTimeUS) {
+  return startWaveformCycles(pin, microsecondsToClockCycles(timeHighUS), microsecondsToClockCycles(timeLowUS), microsecondsToClockCycles(runTimeUS));
+}
+
+int startWaveformCycles(uint8_t pin, uint32_t timeHighCycles, uint32_t timeLowCycles, uint32_t runTimeCycles) {
   if ((pin > 16) || isFlashInterfacePin(pin)) {
     return false;
   }
   Waveform *wave = &waveform[pin];
-  // Adjust to shave off some of the IRQ time, approximately
-  wave->nextTimeHighCycles = microsecondsToClockCycles(timeHighUS);
-  wave->nextTimeLowCycles = microsecondsToClockCycles(timeLowUS);
-  wave->expiryCycle = runTimeUS ? GetCycleCount() + microsecondsToClockCycles(runTimeUS) : 0;
-  if (runTimeUS && !wave->expiryCycle) {
+
+  wave->expiryCycle = runTimeCycles ? GetCycleCount() + runTimeCycles : 0;
+  if (runTimeCycles && !wave->expiryCycle) {
     wave->expiryCycle = 1; // expiryCycle==0 means no timeout, so avoid setting it
   }
 
   uint32_t mask = 1<<pin;
-  if (!(waveformEnabled & mask)) {
+  if (waveformEnabled & mask) {
+    esp8266::InterruptLock il; // Make the variable sets atomic
+    wave->gotoTimeLowCycles = timeLowCycles;
+    wave->gotoTimeHighCycles = timeHighCycles;
```

Comment:
If waveformEnabled is set from timer1 cb, should the lock also cover condition check itself?
(i.e. are we safe while doing & operation?)

Reply:
That's a good question. Usually in a case like this it's pass/fail (i. e. bool result) so it doesn't matter, because you can't corrupt a bool. Worst case is you miss the condition in this pass, which doesn't matter if the timing is so tight, and it'll be caught correctly in the next pass. 
However, on our esp I believe the operation is done in sw, so I guess the answer depends on whether the sw operation is guaranteed to be stable or not when faced with the value changing in the middle of it.

Reply:
@mcspr did find a race condition.  You could do the `if (a&mask)`, take an interrupt stopping the specific pin, and then return to set the next cycle's numbers...which will be ignored because the timeout hit.

It's a corner-case, though, and only when a Tone with a period is specified but is overwritten before it ends.  I should also update the remaining run length here, too.

Reply:
Ah, sorry. But does lock work at all? Looking at initTimer(), timer1 is set as NMI (== meaning we can't stop / hide from it?)

re previous question, since I have a very basic understanding of it. In case of normal isr, [xtensa ISA doc that i found (ref 3.5.4.1, 3.5.4.2 about fetch, 4.4.5.4 about checkInterrupts)](https://0x04.net/~mwk/doc/xtensa.pdf) describes interrupt trigger as beginning of instruction cycle. Does it mean that literal asm lines from `objdump -S` / `g++ -S` location can be used to point to possible 'stop' locations?

Reply:
I was going to complain about the same fundamental issue here, @mcspr, there's multiple mention of NMI in the waveform code, and given that
```
for (int o = 0; o < 8; ++o) {
	tone(D5, o % 2 ? 440 * 4 : 440 / 2);
	yield();
	{
		esp8266::InterruptLock il;
		delayMicroseconds(500000);
	}
}
noTone(D5);
```
produces sounds just fine, I think this PR is broken. Besides, I don't get the point of duplicating my (and @Tech-TX's) extensive work in PR #7022 :-(

Reply:
@dok-net the point of this pr is to provide an immediate fix on top of the current code, which passes @Tech-TX 's checks. This PR shows an almost perfect graph at 40 KHz, with only the extreme high and low having plateaus. @Tech-TX could you please add your linearity graph capture to the PR's main thread?
This is meant to get an interim fix done in time for 2.7.0. It is not meant to replace #7022.